rts/Rendering/IconHandler.cpp

```diff
@@ -90,34 +96,49 @@ bool CIconHandler::AddIcon(
 	const std::string& texName,
 	float size,
 	float distance,
-	bool radAdj
+	bool radAdj,
+	float u0, float v0, float u1, float v1
 ) {
 	if (numIcons == iconData.size()) {
 		LOG_L(L_DEBUG, "[IconHandler::%s] too many icons added (maximum=%u)", __func__, numIcons);
 		return false;
 	}
 
 	unsigned int texID = 0;
-	unsigned int xsize = 0;
-	unsigned int ysize = 0;
 
 	bool ownTexture = true;
 
 	try {
-		CBitmap bitmap;
 
-		if ((ownTexture = !texName.empty() && bitmap.Load(texName))) {
-			texID = bitmap.CreateMipMapTexture();
+		ownTexture = false;
+		if (!texName.empty()) {
+			const auto it = iconTexNameToTexID.find(texName);
 
-			glBindTexture(GL_TEXTURE_2D, texID);
-			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
-			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
-			xsize = bitmap.xsize;
-			ysize = bitmap.ysize;
-		} else {
+			if (it != iconTexNameToTexID.end()){ // Assign it if we found it
+				texID = it->second;
+			}
+			else {
+				CBitmap bitmap;
+				if (bitmap.Load(texName)) {
+					if (FileSystem::GetExtension(texName) == "dds")
+						texID = bitmap.CreateTexture();
+					else
+						texID = bitmap.CreateMipMapTexture();
+
+					iconTexNameToTexID[texName] = texID;
+
+					glBindTexture(GL_TEXTURE_2D, texID);
+					glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+					glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+				}
+				else {
+					// failed to load it, 
+					texID = GetDefaultTexture();
+				}
+			}
+		}
+		else {
```

Comment:
this try-catch block, with its many branches consisting just of `{ texID = bla }`, sounds like an excellent thing to extract into a function where they would become early `return bla`s